### PR TITLE
Case insensitivity of sheet, column, and setting names; plus more performance improvements

### LINKED ISSUE #572: Survey, Choices, Settings should not be case sensistive
If you have workbook with `Survey` or `Choices` rather than `survey` and `choices`, XLSForm reports "Error: The survey sheet is either empty or missing important column headers. I think we these sheet names should be case-insensitive. I thought from #141 addressed it, but using the attached [case.xlsx](https://github.com/XLSForm/pyxform/files/7575768/case.xlsx)
 in pyxform 1.7.0 fails.

### LINKED ISSUE #731: Add error messages if header is missing on survey or choices sheet
A user recently forgot the header in the choices sheet: https://forum.getodk.org/t/enketo-error-when-trying-to-validate-form/50195?u=ln

I think this can happen when copying and pasting across forms. Seems like a reasonable quality of life improvement even if the case is not very common.

### LINKED ISSUE #738: form_id and form_title should not be case-sensitive
If you have a form with Form_ID=foo and Form_Title=bar, those values will not passed through. form_id=foo and form_title=bar will work fine though.

It'd be good to check to make sure all headers are indeed case-sensitive.

Related: https://github.com/XLSForm/pyxform/issues/138

### PR DESCRIPTION
**Pull request by lindsay-stevens**
Closes #738
Closes #731
Closes #572

#### Why is this the best possible solution? Were any other approaches considered?

The ambition is to fix various sheet name and header case-insensitivity / processing issues by:

- standardising the output data structure from all xls2json_backends processors (and input data to workbook_to_json) on `DefinitionData` which has the sheet headers, sheet data, and original sheet names (for spell check).
- using the headers passed in from the backends, so they can be processed once only (rather than repeatedly as before), and then mapping original headers to processed headers.
  - it might make sense to go one step further and require that backends pass the raw content as `list[list[str]]` rather than `list[dict]`. The xls2json_backends are converting `list[list[str]]` to `list[dict]`, only for those `list[dict]` to be re-keyed as a new `list[dict]` with processed headers via xls2json.
- using the SurveyElement class attributes to determine which columns to process and which to leave as-is. For example extra columns in the `choices` sheet need to be verbatim in order for choice_filter expressions to work.
- adding relatively exhaustive tests for:
  - each XLSForm input type with all supported sheet names and a sample of headers
  - all the possibilities for header splitting, normalisation, and nesting

Test cases that justify closing issues:

- #738 
  - test_sheet_columns.TestSettingsColumns.test_case_insensitive_form_id_form_title specifically form_id and form_title
  - test_xls2json_backends.TestXLS2JSONBackends.test_case_insensitivity checks all sheets, including these settings headers
- #731
  - test_sheet_columns.TestSurveyColumns.test_missing_survey_headers error would say that "type" column missing
  - test_sheet_columns.TestChoicesColumns.test_missing_choice_headers error would say that "name" column missing
- #572
  - test_xls2json_backends.TestXLS2JSONBackends.test_case_insensitivity checks all sheet names

Also includes a bunch of performance improvements and/or tidying found along the way, detailed in commit messages. These should help all types of XLSForms, but in particular ones a) with labels or hints, or b) lots of choice lists, or c) written in markdown.

#### What are the regression risks?

If someone is using workbook_to_json directly, or passing in dict to `xls2xform.convert`, then they'll now need to conform to the DefinitionData structure. This could be as simple as `DefinitionDict(**workbook_dict)` or might require specifying the sheet headers.

#### Does this change require updates to documentation? If so, please file an issue [here](https://github.com/XLSForm/xlsform.github.io) and include the link below.

No

#### Before submitting this PR, please make sure you have:
- [x] included test cases for core behavior and edge cases in `tests`
- [x] run `python -m unittest` and verified all tests pass
- [x] run `ruff format pyxform tests` and `ruff check pyxform tests` to lint code
- [x] verified that any code or assets from external sources are properly credited in comments

### CODE REVIEW

**Comment by lognaturel** [comment]
@rgraber @ukanga this PR does make a few more changes to APIs you may be using. From what you've shared, we believe these will take minimal work to adapt to but want to double check that. We would release them as another major version change and then should be able to minimize breaking changes of this kind for a bit.

If there are APIs critical to your use, please consider contributing a few tests on those so there's a record of them being in use. Thanks!

**Review by lognaturel — COMMENTED** [review]
Couple of questions inline.

I went through this relatively quickly because it's a lot. I focused on trying to see changes that might not be adequately covered by tests but nothing jumped out at me. I also verified that the commit messages provided enough context that we could remember what prompted changes if we need to revisit anything.

I'll let all of this information work through my brain and do another final pass.

Since we haven't heard any objections from @ukanga or @rgraber, I think we're ok to merge.

File: pyxform/aliases.py [hunk 1]
```diff
@@ -44,77 +44,57 @@
...
-    "relevant": "bind::relevant",
+    "sms_field": constants.SMS_FIELD,
+    "sms_option": constants.SMS_OPTION,
+    "sms_separator": constants.SMS_SEPARATOR,
+    "sms_allow_media": constants.SMS_ALLOW_MEDIA,
+    "sms_date_format": constants.SMS_DATE_FORMAT,
+    "sms_datetime_format": constants.SMS_DATETIME_FORMAT,
+    "sms_response": constants.SMS_RESPONSE,
+    "compact_tag": ("instance", "odk:tag"),  # used for compact representation
+    "read_only": ("bind", "readonly"),
```
[the remark this answers is not in the document]
Comment (lindsay-stevens) [reply pyxform/aliases.py:63]: Yes that's right - a user can provide the header name as a) the short alias if there is one e.g. "image", b) the full name with "::" delimiter e.g. "media::image", or c) the full name with the deprecated ":" delimiter e.g. "media:image" (except for the "jr:" prefix which is not split off). If using full names, the user needs to choose between doing b) or c) for a given sheet - scenarios for that are in `test_dealias_and_group_headers__use_double_colon_modes` (this is old behaviour but it was selected at the workbook level rather than sheet level in `workbook_to_json`).

Header processing also now tolerates mixed case and spaces before, after, or between header name words (e.g. "constraint_message" could be written as "Constraint Message") - scenarios for that in `TestHeaderProcessing`.

In the aliases module, the full names of aliased headers are pre-split on the delimiter (e.g. `"media::image"` to `("media", "image")`) since the split form is what's used for header processing. This saves pyxform re-splitting these strings for every XLSForm conversion run where the header alias is used (i.e. parse string + create 2 strings + create 1 new tuple).

**Comment by lindsay-stevens** [comment]
Rebased on master to resolve merge conflict in `survey_element.py` from #755. No actual changes, both PRs were writing new lines to the same place in the file.

**Review by lognaturel — APPROVED** [review]
Thanks!
